### PR TITLE
Feature: [Form] Toggle Displaying Percent Type Symbol

### PR DESCRIPTION
**Pull request by d42ohpaz**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #28796
| License       | MIT
| Doc PR        | symfony/symfony-docs#10464 

### PercentType `symbol` option
As of this writing, Symfony will forcibly append a percentage sign (`%`) to all input fields that are of the `PercentType` form type. This PR will introduce a boolean flag called `symbol` that, when `false`, will not display the percentage sign. Each of the default layouts that define `percent_widget` will respect this option.

By default, this new option will be set to `true` so that it maintains backward compatibility. The unit tests have been updated where appropriate, and a new unit test has been added (as appropriate).

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md [hunk 1]
```diff
@@ -19,6 +19,7 @@ CHANGELOG
  * Deprecated the `--no-debug` console option, set the "APP_DEBUG" 
    environment variable to "0" instead.
  * Deprecated the `Templating\Helper\TranslatorHelper::transChoice()` method, use the `trans()` one instead with a `%count%` parameter
+ * Only display `PercentType` percent symbol if the `symbol` option is `true`
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: I don't think we need a changelog entry for this in the FrameworkBundle.

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig [hunk 2]
```diff
@@ -106,12 +106,16 @@
 {%- endblock dateinterval_widget %}
 
 {% block percent_widget -%}
+    {% if symbol == true %}
```
Comment (xabbuh) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: let's shorten this: `{% if symbol %}` (and everywhere else below too)

Comment (xabbuh) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: It also seems like we need to take whitespace control (`i.e.` the `-` character) into account here too.

File: src/Symfony/Bridge/Twig/CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,7 @@ CHANGELOG
  * add bundle name suggestion on wrongly overridden templates paths
  * added `name` argument in `debug:twig` command and changed `filter` argument as `--filter` option
  * deprecated the `transchoice` tag and filter, use the `trans` ones instead with a `%count%` parameter
+ * Update `percent_widget` to consider `symbol` option for `PercentType` when displaying the percent symbol
```
Comment (xabbuh) [inline src/Symfony/Bridge/Twig/CHANGELOG.md]: I don't think we need such an entry in the changelog file of the Twig bridge

File: src/Symfony/Component/Form/CHANGELOG.md [hunk 4]
```diff
@@ -9,6 +9,7 @@ CHANGELOG
  * deprecated calling `FormRenderer::searchAndRenderBlock` for fields which were already rendered
  * added a cause when a CSRF error has occurred
  * deprecated the `scale` option of the `IntegerType`
+ * add `symbol` to `PercentType` options and make available as `FormView` vars
```
Comment (xabbuh) [inline src/Symfony/Component/Form/CHANGELOG.md]: I would reword this a bit:

> added a `symbol` option to the `PercentType` that allows to disable the output of the percent character

File: src/Symfony/Component/Form/Extension/Core/Type/PercentType.php [hunk 5]
```diff
@@ -26,13 +28,22 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+    }
+
     /**
      * {@inheritdoc}
      */
     public function configureOptions(OptionsResolver $resolver)
     {
         $resolver->setDefaults(array(
             'scale' => 0,
+            'symbol' => true,
```
Comment (xabbuh) [inline src/Symfony/Component/Form/Extension/Core/Type/PercentType.php]: we should restrict the allowed types to `bool`:

```php
$resolver->setAllowedTypes('symbol', 'bool');
```

File: src/Symfony/Component/Form/Extension/Core/Type/PercentType.php [hunk 6]
```diff
@@ -26,13 +28,22 @@ public function buildForm(FormBuilderInterface $builder, array $options)
         $builder->addViewTransformer(new PercentToLocalizedStringTransformer($options['scale'], $options['type']));
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    public function buildView(FormView $view, FormInterface $form, array $options)
+    {
+        $view->vars['symbol'] = isset($options['symbol']) ? $options['symbol'] : true;
```
Comment (xabbuh) [inline src/Symfony/Component/Form/Extension/Core/Type/PercentType.php]: Can be shortened as the option will always be set:

```php
$view->vars['symbol'] = $options['symbol'];
```

**Review by xabbuh — CHANGES_REQUESTED** [review]
This looks quite good so far. I have just left a few comments.

Can we think of use cases where you would want to display another symbol than the percent character (then it could make sense to allow arbitrary strings for this option)?

**Comment by dohpaz42** [comment]
> Can we think of use cases where you would want to display another symbol than the percent character (then it could make sense to allow arbitrary strings for this option)?

I missed finding this yesterday, but apparently there are a few [different versions of a percent symbol](https://en.wikipedia.org/wiki/Percent_sign#Encodings):

> U+0025 % Percent sign (HTML &#37;) (&percnt; in HTML5[24]),
> U+2030 ‰ per mille sign (HTML &#8240; · &permil;),
> U+2031 ‱ per ten thousand sign (HTML &#8241;) a.k.a. basis point, and
> U+FF05 ％ fullwidth percent sign (HTML &#65285; · see fullwidth forms)
> U+FE6A ﹪ small percent sign (HTML &#65130; · see Small Form Variants)
> There is also U+066A ٪ ARABIC PERCENT SIGN (HTML &#1642;), which has the circles replaced by square dots set on edge, the shape of the digit 0 in Arabic numerals.

And there are [icons](https://fontawesome.com/v4.7.0/icon/percent). Unless others object, I'll update the PR later today to allow arbitrary text or boolean false.

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig [hunk 7]
```diff
@@ -106,12 +106,16 @@
...
 {% block percent_widget -%}
+    {%- if symbol == true -%}
     <div class="input-group">
         {{- block('form_widget_simple') -}}
         <div class="input-group-append">
             <span class="input-group-text">%</span>
         </div>
     </div>
+    {%- else -%}
+    {{- block('form_widget_simple') -}}
```
Comment (OskarStark) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: could you please add more indention?

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig [hunk 8]
```diff
@@ -106,12 +106,16 @@
 {%- endblock dateinterval_widget %}
 
 {% block percent_widget -%}
+    {%- if symbol == true -%}
     <div class="input-group">
```
Comment (OskarStark) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: could you please add more indention?

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig (hunk 7 above)
Comment (OskarStark) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig]: Same for the other twig parts

**Review by OskarStark — CHANGES_REQUESTED** [review]

**Comment by dohpaz42** [comment]
As for the failed unit tests, I could use some guidance on how to set up an environment and/or execute the tests so that I can reproduce their failures.

**Comment by xabbuh** [comment]
@dohpaz42 Sorry, I missed your question. I think https://symfony.com/doc/current/contributing/code/tests.html should provide some useful insights.

**Comment by xabbuh** [comment]
@9ae8sdf76 Do you think you will be able to finish this pull request?

**Comment by fabpot** [comment]
@xabbuh Can you take over this one?

**Comment by OskarStark** [comment]
I could overtake this too, I just need to know what’s featurewise missing here?

**Comment by fabpot** [comment]
@OskarStark Thank you. I think it's ok feature-wise. It's more about rebasing/squashing and fixing tests if needed after that.

**Comment by OskarStark** [comment]
Done in #30433 

This one can be closed
